light/beacon/binary_merkle.go

```diff
@@ -0,0 +1,555 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package beacon
+
+import (
+	"encoding/binary"
+	"errors"
+	"math/bits"
+	"reflect"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/minio/sha256-simd"
+)
+
+// MerkleValue represents either a 32 byte value or hash node in a binary merkle tree/partial proof
+type (
+	MerkleValue  [32]byte
+	MerkleValues []MerkleValue
+)
+
+var (
+	MerkleValueT = reflect.TypeOf(MerkleValue{})
+	merkleZero   [64]MerkleValue
+)
+
+func init() {
+	hasher := sha256.New()
+	for i := 1; i < 64; i++ {
+		hasher.Reset()
+		hasher.Write(merkleZero[i-1][:])
+		hasher.Write(merkleZero[i-1][:])
+		hasher.Sum(merkleZero[i][:0])
+	}
+}
+
+// UnmarshalJSON parses a merkle value in hex syntax.
+func (m *MerkleValue) UnmarshalJSON(input []byte) error {
+	return hexutil.UnmarshalFixedJSON(MerkleValueT, input, m[:])
+}
+
+// VerifySingleProof verifies a Merkle proof branch for a single value in a
+// binary Merkle tree
+// Note: index is a generalized tree index; bottomLevel is optional (root being
+// level 0) and if greater than len(proof)-1 then the sibling subtrees closest
+// to the root are not stored and all-zero value subtrees are assumed (makes
+// proofs for 24 bit arrays more efficient).
+func VerifySingleProof(proof MerkleValues, index uint64, value MerkleValue, bottomLevel int) (common.Hash, bool) {
```

Comment:
I am wondering if `bottomLevel` is actually useful. 

Usually if we ask something from fullNode, it will respond a value with attached proof. Will this proof omit zero-values by default?
If no, the zero-values is transmitted over p2p anyway and we can just use them.

Reply:
Indeed, it's not used anywhere now so I removed it. Originally I wanted to optimize historical block proofs in LES but then I realized it's unnecessary overcomplication to use a special proof format in order to save a few hundred bytes. In the latest version I'm using the same multiproof format for all beacon state proofs and single proofs are only used in bootstrap data and committee updates (where `bottomLevel` is not needed).